### PR TITLE
[CAPI] Tizen CAPI Example @open sesame 05/04 17:34

### PR DESCRIPTION
**Pull request by jijoongmoon**
Add Tizen CAPI Example with configuration file

- Add example application for tizen capi which run neural network with
  configuration file

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: Applications/Tizen_CAPI/main.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (C) 2019 Samsung Electronics Co., Ltd. All Rights Reserved.
```
Comment (helloahn) [inline Applications/Tizen_CAPI/main.c:2]: maybe 2019->2020?

Comment (jijoongmoon) [reply Applications/Tizen_CAPI/main.c]: fixed. Thanks.

File: Applications/Tizen_CAPI/main.c [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+#include <nntrainer.h>
+
+int
+main (int argc, char *argv[])
+{
+  int status = ML_ERROR_NONE;
+  ml_nnmodel_h handle = NULL;
+  const char *config_file = "./Tizen_CAPI_config.ini";
+  status = ml_nnmodel_construct_with_conf (config_file, &handle);
```
Comment (again4you) [inline Applications/Tizen_CAPI/main.c:34]: `handle` in `ml_nnmodel_construct_with_conf()` API looks like **out** parameter.
However, according to the header file, it is written as `input` parameter as below.

```
* @param[in] model The NNTrainer Model handler from the given description.
```

Comment (jijoongmoon) [reply Applications/Tizen_CAPI/main.c:34]: right. fixed. Thanks.

File: Applications/Tizen_CAPI/main.c [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    return status;
+  status = ml_nnmodel_compile (handle);
+  if (status != ML_ERROR_NONE)
+    return status;
+  status = ml_nnmodel_train (handle);
+  if (status != ML_ERROR_NONE)
+    return status;
+  status = ml_nnmodel_destruct (handle);
+  if (status != ML_ERROR_NONE)
+    return status;
```
Comment (dongju-chae) [inline Applications/Tizen_CAPI/main.c:45]: Q. I wonder whether each ml capi has an internal resource handling mechanism on failures (i.e., whether users don't need to consider resource handling). Otherwise, it's better to give full examples to release immediate resources on failures (if this is reference codes for users).

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+Model = "model.bin"  	# model path to save / read
+minibatch = 32		# mini batch size
+beta1 = 0.9 		# beta 1 for adam
+beta2 = 0.9999	# beta 2 for adam
+epsilon = 1e-7	# epsilon for adam
+
+[DataSet]
+BufferSize=100
+TrainData="trainingSet.dat"
+ValidData="trainingSet.dat"
```
Comment (dongju-chae) [inline Applications/Tizen_CAPI/Tizen_CAPI_config.ini:25]: Same name, intended?

Comment (myungjoo) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:25]: For unittesting, this should be fine, but it'd be better if separated.

File: Applications/Tizen_CAPI/main.c [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  int status = ML_ERROR_NONE;
+  ml_nnmodel_h handle = NULL;
+  const char *config_file = "./Tizen_CAPI_config.ini";
+  status = ml_nnmodel_construct_with_conf (config_file, &handle);
+  if (status != ML_ERROR_NONE)
+    return status;
+  status = ml_nnmodel_compile (handle);
+  if (status != ML_ERROR_NONE)
+    return status;
+  status = ml_nnmodel_train (handle);
```
Comment (myungjoo) [inline Applications/Tizen_CAPI/main.c:40]: Q1. Can we "retrain" a pretrained model with ```ml_nnmodel_train```?
Q2. Can we dynamically adjust training configurations for ```ml_nnmodel_train```?

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# Network Section : Network
+[Network]
+Type = NeuralNetwork	# Network Type : Regression, KNN, NeuralNetwork
+Layers = inputlayer \
+	 outputlayer	#Layers of Neuralnetwork
+Learning_rate = 0.0001 	# Learning Rate
```
Comment (leemgs) [inline Applications/Tizen_CAPI/Tizen_CAPI_config.ini:6]: FYI, Note: Since Keras 2.2.4, the name will be replaced "Learning_rate" with "lr".

File: Applications/Tizen_CAPI/main.c (hunk 5 above)
Comment (jijoongmoon) [reply Applications/Tizen_CAPI/main.c:40]: Q1. yes. if there is model.bin which holds the weight and parameters and defined in configuration file, then nntrainer read it automatically and start train with that data.
Q2. Currently only support static configuration. That means once configuration is read when it starts, there is no way to change it. However, some other APIs like add_layer, create_model, etc. are going be implemented. and then we could find way to do it dynamically.

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+Decay_rate = 0.96	# for the decay_rate for the decayed learning rate
+Decay_steps = 1000       # decay step for the exponential decayed learning rate
+Epoch = 10		# Epoch
+Optimizer = adam 	# Optimizer : sgd (stochastic gradien decent),
+ 	    		#             adam (Adamtive Moment Estimation)
+Cost = cross  		# Cost(loss) function : msr (mean square root error)
+                        #                       cross ( for cross entropy )
+Weight_Decay = l2norm
+weight_Decay_Lambda = 0.005
```
Comment (leemgs) [inline Applications/Tizen_CAPI/Tizen_CAPI_config.ini:15]: I want you to keep the consistency  by replacing "**w**eight_Declay_Lambda" with "**W**eight_Declay_Lambda".

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini (hunk 6 above)
Comment (jijoongmoon) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:6]: Yes. we could support lr as well. I will add it later pr. Thanks.

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini (hunk 4 above)
Comment (jijoongmoon) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:25]: Yes. like @myungjoo mentioned, it is intended. I will find the way to do it better. may be make separate package.

File: Applications/Tizen_CAPI/main.c (hunk 3 above)
Comment (jijoongmoon) [reply Applications/Tizen_CAPI/main.c:45]: Each API has error handling mechanism and this demonstrates how to use them. Thanks.

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini (hunk 7 above)
Comment (jijoongmoon) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:15]: Yes. Actually, nntrainer is not case sensitive.

File: Applications/Tizen_CAPI/Tizen_CAPI_config.ini (hunk 4 above)
Comment (dongju-chae) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:25]: Hmm, then, is `valSet.dat` not used?

Comment (jijoongmoon) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:25]: yes. valSet.dat is not used. this is just test case to evaluate the training mechanism. If we use training data as validation, this should go nearly 100% accuracy even though it is relatively small data. That means training mechanism is Ok at least. This example is intended to check this.

Comment (dongju-chae) [reply Applications/Tizen_CAPI/Tizen_CAPI_config.ini:25]: Thanks for the clear explanation!
